provisioning/transport/amqp/src/amqp.ts

```diff
@@ -2,14 +2,332 @@
+                    com.microsoft:api-version: <API_VERSION>
+                    com.microsoft:client-version: <CLIENT_VERSION>
+                    ```*/
+                    this._amqpBase.attachSenderLink(linkEndpoint, linkOptions, (err, senderLink) => {
+                      if (err) {
+                        debug('_amqpBase.attachSenderLink failed');
+                        debug(err);
+                        /*Codes_SRS_NODE_PROVISIONING_AMQP_16_009: [The `registrationRequest` method shall call its callback with an error if the transport fails to attach the sender link.]*/
+                        /*Codes_SRS_NODE_PROVISIONING_AMQP_16_019: [The `queryOperationStatus` method shall call its callback with an error if the transport fails to attach the sender link.]*/
+                        this._amqpStateMachine.transition('disconnecting', err, null, callback);
+                      } else {
+                        this._senderLink = senderLink;
+                        this._senderLink.on('error', amqpErrorListener);
+                        this._amqpStateMachine.transition('connected', callback);
+                      }
+                    });
+                  }
+                });
+              }
+            });
+          },
+          /*Codes_SRS_NODE_PROVISIONING_AMQP_16_023: [`cancel` shall detach the sender and receiver links and disconnect the AMQP connection.]*/
+          cancel: (callback) => this._amqpStateMachine.transition('disconnecting', null, null, callback),
+          '*': () => this._amqpStateMachine.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback) => callback(),
+          registrationRequest: (request, callback) => {
+
+            /*Codes_SRS_NODE_PROVISIONING_AMQP_16_005: [The `registrationRequest` method shall send a message on the previously attached sender link with a `correlationId` set to a newly generated UUID and the following application properties:
+            ```
+            iotdps-operation-type: iotdps-register;
+            iotdps-forceRegistration: <true or false>;
+            ```
+            ]*/
+            let requestMessage = new AmqpMessage();
+            requestMessage.body = '';
+            requestMessage.applicationProperties = {};
+            requestMessage.properties = {};
+            requestMessage.applicationProperties[MessagePropertyNames.OperationType] = DeviceOperations.Register;
+            requestMessage.applicationProperties[MessagePropertyNames.ForceRegistration] = !!request.forceRegistration;
+            requestMessage.properties.correlationId = uuid.v4();
+
+            debug('initial registration request: ' + JSON.stringify(requestMessage));
+            this._operations[requestMessage.properties.correlationId] = callback;
+            this._senderLink.send(requestMessage, (err) => {
+              if (err) {
+                this._operations[requestMessage.properties.correlationId] = undefined;
```

Comment:
>this._operations[requestMessage.properties.correlationId] = undefined; [](start = 16, length = 70)

IIRC,. I think you want delete this._operations[id].  Otherwise the member is still in the object, but it's set to undefined.

Reply:
that is correct - delete has a slight performance hit but we might not care. it doesn't make a difference for how we use it though. if you think it is cleaner i'm happy to oblige.

---
In reply to: [155602005](https://github.com/Azure/azure-iot-sdk-node/pull/174#discussion_r155602005) [](ancestors = 155602005)